### PR TITLE
fix: Several Fixes for Scroll Handling and Restoration

### PR DESCRIPTION
**Pull request by blainekasten**
## Description

Scroll handling and restoration within Gatsby is diverged from browser basics. There are a few basic things I've found while researching that Gatsby does not do the same (and this PR fixes)

1. url's with encoded hashes do not scroll to matching id's. (e.g., `<h1 id="áccentuated">Scroll Here</h1>`
1. url hashes are not scrolled to when navigating back (POP action)
1. url hashes are not navigated to on page load

While exploring this I also decided to clean up our code a bunch and remove some dependencies. Our usage of ScrollBehavior did not seem necessary. Maybe this is a foot gun and there were good reasons for it to exist.

I also added a `useScrollRestoration` hook that mirrors what Remix was selling for their scroll restoration inside of overflow containers.

I was also able to convert this all to TS.

### Documentation

Maybe documentation to add here?

## Related Issues

Fixes:

shouldUpdateScroll componentDidMount - https://github.com/gatsbyjs/gatsby/issues/23792 (fixed) (breaking?)
accentuated doclinks                 - https://github.com/gatsbyjs/gatsby/issues/23690 (fixed)
POP does not restore scrollPosition  - https://github.com/gatsbyjs/gatsby/issues/8019  (fixed)
doclinks not scrolled to on load     - https://github.com/gatsbyjs/gatsby/issues/21120 (fixed)

## Additional Discussion

I think it would be worth considering, and exploring removing the `shouldUpdateScroll` API from gatsby-browser. In my research, it was only used in user land to patch our bugs.

### CODE REVIEW

**Comment by blainekasten** [comment]
A related question to this is that I don't know if we should let `shouldUpdateScroll` prevent scrolling to hashes. My inclination is to say no, but I don't know if that is a breaking change.

File: packages/gatsby-react-router-scroll/src/scroll-handler.tsx [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+      this.scrollToHash(decodeURI(hash), prevProps)
+    }
+  }
+
+  windowScroll = (
+    position: number,
+    prevProps: LocationContext | undefined
+  ): void => {
+    if (this.shouldUpdateScroll(prevProps, this.props)) {
+      window.scroll(0, position)
```
[the remark this answers is not in the document]
Comment (blainekasten) [reply packages/gatsby-react-router-scroll/src/scroll-handler.tsx:76]: `scrollTo` doesn't exist in IE. So not sure, I could do a scrollTo || scroll

File: packages/gatsby-react-router-scroll/src/scroll-container.tsx [hunk 2]
```diff
@@ -0,0 +1,65 @@
+import React from "react"
```
[the remark this answers is not in the document]
Comment (blainekasten) [reply packages/gatsby-react-router-scroll/src/scroll-container.tsx]: I think it's just for consumer usage. But once we support hooks, the hook approach should be the default. So yeah

File: packages/gatsby-react-router-scroll/src/scroll-handler.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (wardpeet) [reply packages/gatsby-react-router-scroll/src/scroll-handler.tsx:76]: image
MDN says it does and I tested in ie11 and it's there

File: packages/gatsby-react-router-scroll/src/scroll-container.tsx [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+  IPropsWithContextAndLocation
+> {
+  componentDidMount(): void {
+    // eslint-disable-next-line react/no-find-dom-node
+    const node = ReactDOM.findDOMNode(this) as Element
+    const { location, scrollKey } = this.props
+
+    if (!node) return
+
+    node.addEventListener(`scroll`, () => {
```
[the remark this answers is not in the document]
Comment (blainekasten) [reply packages/gatsby-react-router-scroll/src/scroll-container.tsx:36]: Actually no! We are adding an event listener to a node, when the node is removed from the DOM the event listener will be destroyed too.

[the remark this answers is not in the document]
Comment (wardpeet) [reply packages/gatsby-react-router-scroll/src/scroll-container.tsx:36]: oh yeah that's true 👍

File: packages/gatsby-react-router-scroll/src/scroll-handler.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (blainekasten) [reply packages/gatsby-react-router-scroll/src/scroll-handler.tsx]: oh you're right. I looked at the wrong thing :) SWEET!

File: packages/gatsby-react-router-scroll/src/scroll-container.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (blainekasten) [reply packages/gatsby-react-router-scroll/src/scroll-container.tsx]: I upped it one. I added a deprecation warning

image

File: packages/gatsby-react-router-scroll/src/scroll-handler.tsx [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+    }
+
+    if (scrollPosition) {
+      this.windowScroll(scrollPosition, prevProps)
+    } else if (hash) {
+      this.scrollToHash(decodeURI(hash), prevProps)
+    }
+  }
+
+  windowScroll = (
```
[the remark this answers is not in the document]
Comment (blainekasten) [reply packages/gatsby-react-router-scroll/src/scroll-handler.tsx:71]: not sure how? but also it's pretty terse. Seems okay to leave?

File: packages/gatsby-react-router-scroll/src/session-storage.ts [hunk 5]
```diff
@@ -1,13 +1,14 @@
+import { Location } from "history"
 const STATE_KEY_PREFIX = `@@scroll|`
 const GATSBY_ROUTER_SCROLL_STATE = `___GATSBY_REACT_ROUTER_SCROLL`
 
-export default class SessionStorage {
-  read(location, key) {
+export class SessionStorage {
```
[the remark this answers is not in the document]
Comment (blainekasten) [reply packages/gatsby-react-router-scroll/src/session-storage.ts:5]: I think we maybe pass this to gatsby-browser functions. Something to consider

File: packages/gatsby-react-router-scroll/src/scroll-handler.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (wardpeet) [reply packages/gatsby-react-router-scroll/src/scroll-handler.tsx:71]: sure

**Comment by blainekasten** [comment]
@wardpeet wanna give it another test? I found a bug that I fixed. It seems to be working for me now!

File: packages/gatsby-react-router-scroll/src/scroll-handler.tsx [hunk 6]
```diff
@@ -73,7 +73,7 @@ export class ScrollHandler extends React.Component<
     prevProps: LocationContext | undefined
   ): void => {
     if (this.shouldUpdateScroll(prevProps, this.props)) {
-      window.scroll(0, position)
+      ;(window.scrollTo || window.scroll)(0, position)
```
Comment (wardpeet) [inline packages/gatsby-react-router-scroll/src/scroll-handler.tsx]: scrollTo is working on all browsers so we shouldn't do this dance. It's just wanted bytes
```suggestion
      window.scrollTo(0, position)
```

File: packages/gatsby-react-router-scroll/src/session-storage.ts [hunk 7]
```diff
@@ -3,7 +3,7 @@ const STATE_KEY_PREFIX = `@@scroll|`
 const GATSBY_ROUTER_SCROLL_STATE = `___GATSBY_REACT_ROUTER_SCROLL`
 
 export class SessionStorage {
-  read(location: Location, key: string): number {
+  read(location: Location, key: string): number | null {
```
Comment (wardpeet) [inline packages/gatsby-react-router-scroll/src/session-storage.ts:6]: where does null come from?

File: packages/gatsby-react-router-scroll/src/scroll-container.tsx [hunk 8]
```diff
@@ -23,9 +25,25 @@ interface IPropsWithContextAndLocation extends IProps {
...
 
+let hasNotWarnedDeprecation = true
+
 class ScrollContainerImplementation extends React.Component<
   IPropsWithContextAndLocation
 > {
+  constructor(props) {
+    super(props)
+
+    if (hasNotWarnedDeprecation) {
```
Comment (wardpeet) [inline packages/gatsby-react-router-scroll/src/scroll-container.tsx:36]: Only show deprecations on dev mode
```suggestion
    if (process.env.NODE_ENV !== 'production' && hasNotWarnedDeprecation) {
```

File: packages/gatsby-react-router-scroll/src/session-storage.ts (hunk 7 above)
Comment (blainekasten) [reply packages/gatsby-react-router-scroll/src/session-storage.ts]: huh, i guess it was left over from some development. removing

**Comment by blainekasten** [comment]
@wardpeet finally came back around to this and fixed everything up. Think it should be good to ship now! Take another review?

**Comment by josephmarkus** [comment]
@blainekasten @wardpeet when is this expected to be released? Also, what's the last version that didn't have this change?

**Comment by coreyward** [comment]
Heads up: this doesn't appear to have made it into the [changelog](https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby/CHANGELOG.md), so it's hard to see which version of Gatsby this is fixed for.

**Comment by blainekasten** [comment]
@coreyward Fixed this in the above commit. Thanks for the call out. It was fixed in 2.23.5!

**Comment by blimpmason** [comment]
@wardpeet @blainekasten I'm not 100% sure but I suspect that API changes included in this merged PR have introduced issues with shouldUpdateScroll browser API. It also seems the browser API documentation on `getSavedScrollPosition` is outdated or incorrect (see https://github.com/gatsbyjs/gatsby/issues/23842#issuecomment-632970403). 

It appears the scroll position is no longer saved in session storage as an [x, y] array, but rather as a single number for y position. All the Gatsby sites I've built with with animated page transitions no longer successfully restore scroll position on pop state — instead the user is always returned to the top of the page on browser back/forward. Up till now I've use this code snippet in `gatsby-browser.js` with no issue:  

```
const transitionDelay = 500;

export const shouldUpdateScroll = ({
  routerProps: { location },
  getSavedScrollPosition
}) => {
  if (location.action === "PUSH") {
    window.setTimeout(() => window.scrollTo(0, 0), transitionDelay);
  } else {
    const savedPosition = getSavedScrollPosition(location);
    window.setTimeout(
      () => window.scrollTo(...(savedPosition || [0, 0])),
      transitionDelay
    );
  }
  return false;
};
```

After digging it appears the getSavedScrollPosition now requires both location and location.key as arguments, and returns a single number rather than x/y coordinates. So I've updated to this:

```
const transitionDelay = 500;

export const shouldUpdateScroll = ({
  routerProps: { location },
  getSavedScrollPosition
}) => {
  if (location.action === "PUSH") {
    window.setTimeout(() => window.scrollTo(0, 0), transitionDelay);
  } else {
    const savedYPosition = getSavedScrollPosition(location);
    window.setTimeout(
      () => window.scrollTo(0, savedYPosition || 0)),
      transitionDelay
    );
  }
  return false;
};
```

But unfortunately that update doesn't solve the issue either. It also seems that the scroll position is logged at a different time in the route change process because by the time it's called above (even with setTimeout delay matching the page transition speed), the savedYPosition comes back as `0` every single time. 

This simple gatsby-browser trick used to work every time, as long as the setTimeout delay matches the speed of the page transition. Now it never seems to work. I couldn't find anything in the changelog that shed any light on the issue but found this PR as I browsed through the history of the files related to scroll handling and restoration.

**Comment by blimpmason** [comment]
A little more info — I was able to get the scroll restoration to work on popstate using the following workaround code that accesses the session storage directly instead of using `getSavedScrollPosition`, but it updates scroll position immediately when the page transition starts instead of waiting for the setTimeout. The UX is quite jerky and defeats the purpose of having a smooth page transition. That's what makes me think there's something different and breaking about when `shouldUpdateScroll` fires or whether it respects the setTimeout delay:

```
const getLatestSavedScrollPosition = ({ pathname, key }) => {
  const fullKey = `@@scroll|${pathname}|${key}`;
  return sessionStorage.getItem(fullKey);
};

export const shouldUpdateScroll = ({ routerProps: { location } }) => {
  console.log("shouldUpdateScroll");
  if (location.action === "PUSH") {
    window.setTimeout(() => window.scrollTo(0, 0), 600);
  } else {
    const yPosition = getLatestSavedScrollPosition(location);
    window.setTimeout(() => window.scrollTo(0, yPosition || 0), 600);
  }
  return false;
};
```

**Comment by wardpeet** [comment]
@blimpmason can you open an issue cause merged Prs are hardly checked

**Pull request merged**